File: Python/131.palindrome-partitioning.py
#
# @lc app=leetcode id=131 lang=python
#
# [131] Palindrome Partitioning
#
# https://leetcode.com/problems/palindrome-partitioning/description/
#
# algorithms
# Medium (43.30%)
# Likes:    1221
# Dislikes: 49
# Total Accepted:    188.2K
# Total Submissions: 433.7K
# Testcase Example:  '"aab"'
#
# Given a string s, partition s such that every substring of the partition is a
# palindrome.
# 
# Return all possible palindrome partitioning of s.
# 
# Example:
# 
# 
# Input: "aab"
# Output:
# [
# ⁠ ["aa","b"],
# ⁠ ["a","a","b"]
# ]
# 
# 
#

# @lc code=start
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        res = []
        self.dfs(s, [], res)
        return res
    
    def dfs(self, s, path, res):
        if not s:
            res.append(path)
            return 

        for i in range(1, len(s)+1):
            if self.is_palindrome(s[:i]):     
                self.dfs(s[i:], path+[s[:i]], res)
    
    def is_palindrome(self, s):
        return s==s[::-1]
        

# @lc code=end

sol = Solution()
s = 'aba'
print(sol.partition(s))

File: C++/131.palindrome-partitioning.cpp
/*
 * @lc app=leetcode id=131 lang=cpp
 *
 * [131] Palindrome Partitioning
 *
 * https://leetcode.com/problems/palindrome-partitioning/description/
 *
 * algorithms
 * Medium (44.63%)
 * Likes:    1388
 * Dislikes: 54
 * Total Accepted:    201.4K
 * Total Submissions: 451.1K
 * Testcase Example:  '"aab"'
 *
 * Given a string s, partition s such that every substring of the partition is
 * a palindrome.
 * 
 * Return all possible palindrome partitioning of s.
 * 
 * Example:
 * 
 * 
 * Input: "aab"
 * Output:
 * [
 * ⁠ ["aa","b"],
 * ⁠ ["a","a","b"]
 * ]
 * 
 * 
 */
#include<vector>
#include<string>
using namespace std;

// @lc code=start
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> path;
        DFS(s, 0, path, res);
        return res;
    }
    void DFS(string& s, int idx, vector<string>& path, vector<vector<string>>& res) {
        if (idx==s.size()) {
            res.push_back(path);
            return;
        }

        for (int i=idx; i<s.size(); i++) {
            if (isPalindrome(s, idx, i)) {
                path.push_back(s.substr(idx, i-idx+1));
                DFS(s, i+1, path, res);
                path.pop_back();
            }
        }
    }

    bool isPalindrome(string& s, int l, int r) {
        while (l<r) {
            if (s[l++]!=s[r--])
                return false;
        }
        return true;
    }
};
// @lc code=end

